staging/src/k8s.io/apiserver/pkg/server/config.go

```diff
@@ -480,16 +440,14 @@ func (c completedConfig) buildHandlers(s *GenericAPIServer, delegate http.Handle
 	}
 
 	installAPI(s, c.Config)
-	if delegate != nil {
-		s.FallThroughHandler.NotFoundHandler(delegate)
-	} else if c.EnableIndex {
-		s.FallThroughHandler.NotFoundHandler(routes.IndexLister{
+	if delegate == nil && c.EnableIndex {
+		s.Handler.PostGoRestfulMux.NotFoundHandler(routes.IndexLister{
 			StatusCode:   http.StatusNotFound,
 			PathProvider: s.listedPathProvider,
 		})
 	}
 
-	s.Handler = c.BuildHandlerChainFunc(s.HandlerContainer.ServeMux, c.Config)
+	s.Handler.CompleteHandler = c.BuildHandlerChainFunc(s.Handler.CompleteHandler, c.Config)
```

Comment:
> this in-place update is ugly. I would prefer to have an immutable APIServerHandler.

I don't think that's realistic since the contents underneath it change when API groups are added: aggregator, kube-apiserver, and apiextensions-server all mutate and anyone adding their own endpoints (also likely) will mutate the handler.